exonum/src/blockchain/block.rs

```diff
@@ -19,8 +19,57 @@ use crate::{
     crypto::Hash,
     helpers::{Height, ValidatorId},
     messages::{Precommit, Verified},
-    proto,
+    proto::{self, BinaryMap},
 };
+use exonum_merkledb::BinaryValue;
+use std::borrow::Cow;
+
+/// Trait that represents key in block header entry map. Provide
+/// mapping between `NAME` of the entry and its value.
+///
+/// # Usage
+///
+/// ```ignore
+/// use exonum::blockchain::{Block, BlockHeaderKey};
+///
+/// struct SomeData {}
+///
+/// impl BlockHeaderKey for SomeData {
+///    const NAME: &'static str = "data";
+///    type Value = Self;
+/// }
+///
+/// let mut block = Block { ... };
+///
+/// let data = SomeData {};
+/// block.insert::<SomeData>(data);
+/// ```
+pub trait BlockHeaderKey {
+    /// Key name.
+    const NAME: &'static str;
+    /// Type of the value associated with this key.
+    type Value: BinaryValue;
+}
+
+/// Expandable set of entries allowed to be added to the block.
+pub type BlockHeaderEntries = BinaryMap<String, Vec<u8>>;
```

Comment:
Nit: Since we use naming `Block` rather than `BlockHeader`, it could make sense to use a name like `BlockEntries` or `CustomEntries` / `AdditionalEntries` (to prevent [stuttering](https://doc.rust-lang.org/1.0.0/style/style/naming/README.html#avoid-redundant-prefixes-[rfc-356])).

Reply:
I'd even suggest `AdditionalHeaders` by analogy with HTTP headers. IMO `Entry` is as meaningful as `Manager` or `Processor` in the most of contexts.